### PR TITLE
Modify halt_error() to dump received halt structure

### PR DESCRIPTION
**Pull request by orenc17A**

### CODE REVIEW

File: source/main.cpp [hunk 1]
```diff
@@ -37,10 +37,40 @@ static uint32_t get_version(void) {
...
-static void halt_error(int reason) {
-    printf("***** uVisor debug box example *****\n");
-    printf("Tried to access address 0xFFFFFFFF which is not allowed\n");
-	printf("Bye Bye Now!!!!!!\n");
+static void halt_error(THaltError reason, const THaltInfo *halt_info) {
+    const char *exception_stack[] = {"R0", "R1", "R2", "R3", "R12", "LR", "PC", "xPSR"};
+    printf("***** uVisor debug box example *****")
+    printf("Failed with an error code: 0x%08x\r\n\r\n", reason);
+
+	if (halt_info == NULL) {
```
Comment (Patater) [inline source/main.cpp]: Don't indent here. Only indent for new blocks.

File: source/main.cpp [hunk 2]
```diff
@@ -37,10 +37,40 @@ static uint32_t get_version(void) {
...
+		return;
+	}
+
+    if ((halt_info->valid_data & HALT_INFO_STACK_FRAME)) {
+        printf("Exception stack frame:\r\n");
+        for (int i = sizeof(exception_stack) / sizeof(exception_stack[0]) - 1; i >= 0; i--)
+        {
```
Comment (Patater) [inline source/main.cpp]: Place this brace on same line as the `for`.

File: test/log.txt [hunk 3]
```diff
@@ -1,3 +1,2 @@
 ***** uVisor debug box example *****
-Tried to access address 0xFFFFFFFF which is not allowed
-Bye Bye Now!!!!!!
+Failed with an error code:
```
Comment (Patater) [inline test/log.txt]: We should also have the long prints here, including the MMFAR.

File: source/main.cpp [hunk 4]
```diff
@@ -37,10 +37,40 @@ static uint32_t get_version(void) {
     return 0;
 }
 
-static void halt_error(int reason) {
-    printf("***** uVisor debug box example *****\n");
-    printf("Tried to access address 0xFFFFFFFF which is not allowed\n");
-	printf("Bye Bye Now!!!!!!\n");
+static void halt_error(THaltError reason, const THaltInfo *halt_info) {
```
Comment (Patater) [inline source/main.cpp]: The brace at the end of this line should be on a new line below.

**Comment by Patater** [comment]
Please fix all warnings, including the subset below.

```
[Warning] main.cpp@53,112: format '%x' expects argument of type 'unsigned int', but argument 3 has type 'uint32_t {aka long unsigned int}' [-Wformat=]
[Warning] main.cpp@60,52: format '%x' expects argument of type 'unsigned int', but argument 2 has type 'uint32_t {aka long unsigned int}' [-Wformat=]
[Warning] main.cpp@61,54: format '%x' expects argument of type 'unsigned int', but argument 2 has type 'uint32_t {aka long unsigned int}' [-Wformat=]
[Warning] main.cpp@62,57: format '%x' expects argument of type 'unsigned int', but argument 2 has type 'uint32_t {aka long unsigned int}' [-Wformat=]
[Warning] main.cpp@66,52: format '%x' expects argument of type 'unsigned int', but argument 2 has type 'uint32_t {aka long unsigned int}' [-Wformat=]
[Warning] main.cpp@67,52: format '%x' expects argument of type 'unsigned int', but argument 2 has type 'uint32_t {aka long unsigned int}' [-Wformat=]
[Warning] main.cpp@68,52: format '%x' expects argument of type 'unsigned int', but argument 2 has type 'uint32_t {aka long unsigned int}' [-Wformat=]
[Warning] main.cpp@69,52: format '%x' expects argument of type 'unsigned int', but argument 2 has type 'uint32_t {aka long unsigned int}' [-Wformat=]
[Warning] main.cpp@70,53: format '%x' expects argument of type 'unsigned int', but argument 2 has type 'uint32_t {aka long unsigned int}' [-Wformat=]
[Warning] main.cpp@71,52: format '%x' expects argument of type 'unsigned int', but argument 2 has type 'uint32_t {aka long unsigned int}' [-Wformat=]
```

File: source/main.cpp [hunk 5]
```diff
@@ -37,10 +37,40 @@ static uint32_t get_version(void) {
...
+    const char *exception_stack[] = {"R0", "R1", "R2", "R3", "R12", "LR", "PC", "xPSR"};
+    printf("***** uVisor debug box example *****")
```
Comment (Patater) [inline source/main.cpp]: Add semicolon at end of this line.

File: test/log.txt (hunk 3 above)
Comment (orenc17A) [inline test/log.txt]: its printed later

File: source/main.cpp [hunk 6]
```diff
@@ -14,6 +14,8 @@
  * See the License for the specific language governing permissions and
  * limitations under the License.
  */
+
```
Comment (Patater) [inline source/main.cpp]: Don't add newline here.

File: test/log.txt [hunk 7]
```diff
@@ -1,3 +1,25 @@
 ***** uVisor debug box example *****
-Tried to access address 0xFFFFFFFF which is not allowed
-Bye Bye Now!!!!!!
+Failed with an error code\: 0x[0-9a-fA-F]{8}
```
Comment (Patater) [inline test/log.txt]:
```diff
- Failed with an error code\: 0x[0-9a-fA-F]{8}
+ Failed with an error code\: 0xTHEACTUALCODE
```

File: source/main.cpp [hunk 8]
```diff
@@ -33,14 +35,57 @@ uint8_t __attribute__((section(".keep.uvisor.bss.boxes"), aligned(32))) __boxes_
 
 /* Configure box 0 as the debug box. */
 
-static uint32_t get_version(void) {
+static uint32_t get_version(void) 
+{
```
Comment (Patater) [inline source/main.cpp]: Make style changes in a separate commit from semantic changes.

File: source/main.cpp [hunk 9]
```diff
@@ -37,10 +38,52 @@ static uint32_t get_version(void) {
...
+    if (halt_info == NULL) {
+        return;
+    }
+
+    if ((halt_info->valid_data & HALT_INFO_STACK_FRAME)) {
+        printf("Exception stack frame:\r\n");
+        for (int i = sizeof(exception_stack) / sizeof(exception_stack[0]) - 1; i >= 0; i--) {
+            printf("SP[%02d]: 0x%08" PRIx32 " | %s\r\n", i, ((uint32_t *)&halt_info->stack_frame)[i], exception_stack[i]);
+        }
+
```
Comment (Patater) [inline source/main.cpp]: Please don't add trailing whitespace. Configure your editor to remove trailing whitespace from lines you modify.

File: source/main.cpp [hunk 10]
```diff
@@ -33,14 +33,17 @@ uint8_t __attribute__((section(".keep.uvisor.bss.boxes"), aligned(32))) __boxes_
 
 /* Configure box 0 as the debug box. */
 
-static uint32_t get_version(void) {
+static uint32_t get_version(void)
+{
     return 0;
 }
 
 static void halt_error(int reason) {
```
Comment (Patater) [inline source/main.cpp]: Please fix this brace.

File: source/main.cpp [hunk 11]
```diff
@@ -33,14 +33,17 @@ uint8_t __attribute__((section(".keep.uvisor.bss.boxes"), aligned(32))) __boxes_
...
     printf("***** uVisor debug box example *****\n");
     printf("Tried to access address 0xFFFFFFFF which is not allowed\n");
 	printf("Bye Bye Now!!!!!!\n");
+static void halt_error(THaltError reason, const THaltInfo *halt_info)
```
Comment (Patater) [inline source/main.cpp]: Don't add the new halt_error handler here yet. Save that for the "Modify halt_error()..." commit.
